File: packages/coding-agent/src/core/mcp/connection-store.ts
// Local connection records for MCP services. Separate from credentials (auth.json):
// a record captures the verified connection state for a connectionId — the stable
// alias the kernel dispatches through — plus the catalog serviceId it connects and
// the endpoint the verification ran against. Tokens never live here.
//
// Writes use the same file-lock + read-modify-write pattern as auth storage: the
// interactive client and the daemon both mutate this file, so every flush re-reads
// the latest on-disk state under a proper-lockfile lock and applies only this
// instance's pending operations. Verification results apply under a guard so a
// stale probe can never mark a newer grant (or a logged-out connection) verified.

import { existsSync, mkdirSync, readFileSync } from "node:fs";
import { dirname } from "node:path";
import lockfile from "proper-lockfile";
import { realpathIfPresentSync, writeFileAtomicSync } from "../../utils/atomic-file.js";

export type McpConnectionRecordStatus = "connected" | "pending" | "error";

export interface McpConnectionRecord {
	/** Auth.json key suffix (`mcp:<connectionId>`) and the kernel dispatch id. */
	connectionId: string;
	/** Catalog service id; deliberately distinct from connectionId (aliases remain possible). */
	serviceId: string;
	/** Endpoint the record's verification ran against. */
	endpoint: string;
	label: string;
	status: McpConnectionRecordStatus;
	createdAt: number;
	updatedAt: number;
	/** Epoch ms of the last successful handshake + tools/list. */
	verifiedAt?: number;
	toolCount?: number;
	/** Fixed, safe failure category (never URLs or server-controlled text). */
	lastError?: string;
}

interface McpConnectionsFile {
	version: 1;
	connections: Record<string, McpConnectionRecord>;
}

type PendingOp =
	| { kind: "upsert"; record: McpConnectionRecord }
	| { kind: "remove"; connectionId: string }
	| {
			kind: "verify";
			record: McpConnectionRecord;
			/** Evaluated at flush time under the lock; a failed guard discards the result. */
			isStillCurrent: () => boolean;
	  };

const MAX_LAST_ERROR_LENGTH = 500;

function sanitizeRecord(raw: unknown, connectionId: string): McpConnectionRecord | undefined {
	if (!raw || typeof raw !== "object") return undefined;
	const value = raw as Partial<McpConnectionRecord>;
	const string = (input: unknown): string | undefined => (typeof input === "string" ? input : undefined);
	const status = value.status;
	if (
		!string(value.connectionId) ||
		!string(value.serviceId) ||
		!string(value.endpoint) ||
		!string(value.label) ||
		(status !== "connected" && status !== "pending" && status !== "error")
	) {
		return undefined;
	}
	const record: McpConnectionRecord = {
		connectionId: string(value.connectionId) ?? connectionId,
		serviceId: string(value.serviceId) ?? "",
		endpoint: string(value.endpoint) ?? "",
		label: string(value.label) ?? string(value.connectionId) ?? connectionId,
		status,
		createdAt: typeof value.createdAt === "number" ? value.createdAt : Date.now(),
		updatedAt: typeof value.updatedAt === "number" ? value.updatedAt : Date.now(),
	};
	if (typeof value.verifiedAt === "number") record.verifiedAt = value.verifiedAt;
	if (typeof value.toolCount === "number") record.toolCount = value.toolCount;
	const lastError = string(value.lastError);
	if (lastError) record.lastError = lastError.slice(0, MAX_LAST_ERROR_LENGTH);
	return record;
}

function parseRecords(raw: string | undefined): Map<string, McpConnectionRecord> {
	const records = new Map<string, McpConnectionRecord>();
	if (!raw) return records;
	let parsed: unknown;
	try {
		parsed = JSON.parse(raw);
	} catch {
		return records;
	}
	const file = parsed as Partial<McpConnectionsFile>;
	if (!file || file.version !== 1 || typeof file.connections !== "object" || file.connections === null) {
		return records;
	}
	for (const [connectionId, value] of Object.entries(file.connections)) {
		const record = sanitizeRecord(value, connectionId);
		if (record && record.connectionId === connectionId) {
			records.set(connectionId, record);
		}
	}
	return records;
}

function serializeRecords(records: ReadonlyMap<string, McpConnectionRecord>): string {
	const file: McpConnectionsFile = {
		version: 1,
		connections: Object.fromEntries([...records.entries()].sort(([left], [right]) => left.localeCompare(right))),
	};
	return `${JSON.stringify(file, null, "\t")}\n`;
}

/** File-backed store of MCP connection records; corrupt or missing files reset to empty. */
export class McpConnectionStore {
	private recordsById = new Map<string, McpConnectionRecord>();
	private pendingOps: PendingOp[] = [];
	private flushChain: Promise<void> = Promise.resolve();

	private constructor(private readonly path: string) {}

	static open(path: string): McpConnectionStore {
		const store = new McpConnectionStore(path);
		store.load();
		return store;
	}

	/** Re-read the file. Tolerates a missing or corrupt file by resetting to empty. */
	load(): void {
		this.recordsById = parseRecords(
			existsSync(this.path) ? safeReadFileSync(realpathIfPresentSync(this.path)) : undefined,
		);
	}

	get(connectionId: string): McpConnectionRecord | undefined {
		return this.recordsById.get(connectionId);
	}

	records(): readonly McpConnectionRecord[] {
		return [...this.recordsById.values()];
	}

	/** Queue an unconditional upsert; applied in memory immediately and on disk at flush. */
	upsert(record: McpConnectionRecord): void {
		this.applyUpsert(this.recordsById, record);
		this.pendingOps.push({ kind: "upsert", record: { ...record } });
	}

	/** Queue an unconditional removal; applied in memory immediately and on disk at flush. */
	remove(connectionId: string): void {
		this.recordsById.delete(connectionId);
		this.pendingOps.push({ kind: "remove", connectionId });
	}

	/**
	 * Queue a verification result behind a guard. The guard is re-evaluated at flush
	 * time under the file lock, so a result computed against an old grant (or a
	 * connection that has since been disconnected) is discarded instead of
	 * resurrecting a stale record.
	 */
	queueVerifyResult(record: McpConnectionRecord, isStillCurrent: () => boolean): void {
		this.pendingOps.push({ kind: "verify", record: { ...record }, isStillCurrent });
	}

	/**
	 * Persist pending mutations. Serialized within this instance; the file lock
	 * orders us against other processes. Each flush re-reads the latest on-disk
	 * state and applies only this instance's pending operations, so concurrent
	 * writers cannot lose each other's records.
	 */
	flush(): Promise<void> {
		const run = async (): Promise<void> => {
			mkdirSync(dirname(this.path), { recursive: true });
			if (!existsSync(this.path)) {
				writeFileAtomicSync(this.path, serializeRecords(new Map()), { mode: 0o600 });
			}
			let lockCompromised = false;
			let lockCompromisedError: Error | undefined;
			const release = await lockfile.lock(realpathIfPresentSync(this.path), {
				retries: {
					retries: 10,
					factor: 2,
					minTimeout: 100,
					maxTimeout: 10000,
					randomize: true,
				},
				stale: 30000,
				onCompromised: (error) => {
					lockCompromised = true;
					lockCompromisedError = error as Error;
				},
			});
			try {
				if (lockCompromised) throw lockCompromisedError ?? new Error("MCP connection store lock was compromised");
				const records = parseRecords(safeReadFileSync(realpathIfPresentSync(this.path)));
				// Splice only after acquiring the lock: operations queued while we
				// waited belong to this flush, not the previous one.
				const operations = this.pendingOps.splice(0);
				for (const operation of operations) {
					if (operation.kind === "remove") {
						records.delete(operation.connectionId);
					} else if (operation.kind === "upsert") {
						this.applyUpsert(records, operation.record);
					} else if (operation.isStillCurrent()) {
						this.applyUpsert(records, operation.record);
					}
				}
				writeFileAtomicSync(realpathIfPresentSync(this.path), serializeRecords(records), { mode: 0o600 });
				if (lockCompromised) throw lockCompromisedError ?? new Error("MCP connection store lock was compromised");
				this.recordsById = records;
			} finally {
				if (lockCompromised) await release().catch(() => undefined);
				else await release();
			}
		};
		this.flushChain = this.flushChain.then(run, run);
		return this.flushChain;
	}

	private applyUpsert(records: Map<string, McpConnectionRecord>, record: McpConnectionRecord): void {
		const now = Date.now();
		const previous = records.get(record.connectionId);
		const next: McpConnectionRecord = {
			...record,
			createdAt: previous?.createdAt ?? record.createdAt ?? now,
			updatedAt: now,
			lastError: record.lastError ? record.lastError.slice(0, MAX_LAST_ERROR_LENGTH) : undefined,
		};
		if (next.status !== "connected") {
			delete next.verifiedAt;
			delete next.toolCount;
		}
		records.set(record.connectionId, next);
	}
}

function safeReadFileSync(path: string): string | undefined {
	try {
		return readFileSync(path, "utf8");
	} catch {
		return undefined;
	}
}

export type { McpConnectionsFile, PendingOp };

File: crates/pa-core/src/mcp/connection_store.rs
//! Durable MCP connection records (port of
//! `packages/coding-agent/src/core/mcp/connection-store.ts`, the record
//! core): the verified connection state per connectionId — the stable alias
//! the kernel dispatches through — plus the catalog serviceId it connects
//! and the endpoint the verification ran against. Tokens never live here.
//!
//! The record endpoint is the ENDPOINT PIN: an installed connection keeps
//! its approved endpoint for dispatch and management even if the catalog
//! later changes the service URL, and the resolver pins vanished sources
//! from it. Writes are locked read-modify-write (the auth-storage pattern:
//! lockfile + atomic temp-file rename at mode 0600).

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};

use crate::settings::storage::atomic_write;

/// The connections file version.
const FILE_VERSION: u8 = 1;
/// One connection's verified state.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpConnectionRecord {
    /// Auth key suffix (`mcp:<connectionId>`) and the kernel dispatch id.
    pub connection_id: String,
    /// Catalog service id; deliberately distinct from the connectionId
    /// (`acme-2` aliases keep their parent's service id).
    pub service_id: String,
    /// Endpoint the record's verification ran against (the pin).
    pub endpoint: String,
    pub label: String,
    pub status: McpConnectionStatus,
    pub created_at: u64,
    pub updated_at: u64,
    /// Epoch ms of the last successful handshake + tools/list.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub verified_at: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_count: Option<usize>,
    /// Fixed, safe failure category (never URLs or server-controlled text).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
    /// Opaque one-time ownership nonce for a pending login attempt.
    #[serde(rename = "attemptId", default, skip_serializing_if = "Option::is_none")]
    pub attempt_id: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum McpConnectionStatus {
    Connected,
    Pending,
    Error,
}

impl McpConnectionStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            McpConnectionStatus::Connected => "connected",
            McpConnectionStatus::Pending => "pending",
            McpConnectionStatus::Error => "error",
        }
    }
}

/// The connections file: `{ "version": 1, "connections": { id: record } }`.
#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct ConnectionsFile {
    version: u8,
    #[serde(default)]
    connections: HashMap<String, McpConnectionRecord>,
}

/// Durable connection-record store. One instance per process is fine: every
/// mutating op re-reads the latest on-disk state under the lockfile and
/// applies only this call's change, so the interactive client and the daemon
/// worker can both mutate the file.
#[derive(Debug)]
pub struct McpConnectionStore {
    path: PathBuf,
    records: HashMap<String, McpConnectionRecord>,
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|since| since.as_millis() as u64)
        .unwrap_or_default()
}

/// Keep only structurally valid records on load: a corrupt or partial entry
/// is dropped, never served (matching the TS `sanitizeRecord`).
fn record_valid(record: &McpConnectionRecord) -> bool {
    !record.connection_id.is_empty()
        && !record.service_id.is_empty()
        && !record.endpoint.is_empty()
        && !record.label.is_empty()
        && record.created_at > 0
        && record.updated_at > 0
}

impl McpConnectionStore {
    /// Open the store at `<agent-dir>/mcp-connections.json` and load the
    /// latest records from disk. A missing file is an empty store.
    pub fn open(path: impl Into<PathBuf>) -> Self {
        let mut store = Self {
            path: path.into(),
            records: HashMap::new(),
        };
        store.load();
        store
    }

    /// In-memory store (tests, embedded hosts).
    pub fn in_memory() -> Self {
        Self {
            path: PathBuf::new(),
            records: HashMap::new(),
        }
    }

    /// Re-read the latest on-disk state (another process may have written).
    pub fn load(&mut self) {
        if self.path.as_os_str().is_empty() {
            return;
        }
        self.records = read_records(&self.path);
    }

    /// All records, sorted by connectionId (deterministic order).
    pub fn records(&self) -> Vec<McpConnectionRecord> {
        let mut records: Vec<McpConnectionRecord> = self.records.values().cloned().collect();
        records.sort_by(|a, b| a.connection_id.cmp(&b.connection_id));
        records
    }

    /// One record by connectionId.
    pub fn get(&self, connection_id: &str) -> Option<&McpConnectionRecord> {
        self.records.get(connection_id)
    }

    /// Insert or update one record (locked read-modify-write: the on-disk
    /// state wins for every OTHER id first). Install paths and tests that
    /// build records by hand.
    #[cfg(test)]
    pub fn upsert(&mut self, record: McpConnectionRecord) -> Result<(), anyhow::Error> {
        self.with_disk_state(|records| {
            records.insert(record.connection_id.clone(), record.clone());
        })
    }

    /// Remove one record entirely (disconnect keeps nothing behind).
    pub fn remove(&mut self, connection_id: &str) -> Result<(), anyhow::Error> {
        self.with_disk_state(|records| {
            records.remove(connection_id);
        })
    }

    /// Apply one verification outcome under the shared guard: a stale probe
    /// (the credential changed) never marks a newer grant or a logged-out
    /// connection verified — the result is discarded and the caller is
    /// told. A fresh install (no record yet) commits only when the probe's
    /// credential is still the stored one; a vanished record (removed
    /// mid-probe) commits nothing.
    pub fn apply_verify_result(
        &mut self,
        record: McpConnectionRecord,
        still_current: bool,
    ) -> Result<bool, anyhow::Error> {
        let connection_id = record.connection_id.clone();
        let mut committed = false;
        self.with_disk_state(|records| {
            match records.get(&connection_id) {
                None if still_current => {
                    // A fresh install creates the record.
                    records.insert(connection_id.clone(), record.clone());
                    committed = true;
                }
                Some(existing) if existing.attempt_id.is_none() && still_current => {
                    records.insert(connection_id.clone(), record.clone());
                    committed = true;
                }
                _ => {}
            }
        })?;
        Ok(committed)
    }

    /// Locked read-modify-write: reload the disk state, apply the change,
    /// write back atomically. In-memory stores just mutate.
    fn with_disk_state(
        &mut self,
        change: impl FnOnce(&mut HashMap<String, McpConnectionRecord>),
    ) -> Result<(), anyhow::Error> {
        if self.path.as_os_str().is_empty() {
            change(&mut self.records);
            return Ok(());
        }
        let lock = crate::platform::lock_dir::LockDir::acquire(
            &self.path,
            std::time::Duration::from_secs(10),
        )
        .map_err(|error| anyhow::anyhow!("connection store lock: {error}"))?;
        let mut records = read_records(&self.path);
        change(&mut records);
        write_records(&self.path, &records)?;
        drop(lock);
        self.records = records;
        Ok(())
    }
}

/// Read and sanitize the file's records (missing file = empty).
fn read_records(path: &Path) -> HashMap<String, McpConnectionRecord> {
    let Ok(content) = std::fs::read_to_string(path) else {
        return HashMap::new();
    };
    let Ok(file) = serde_json::from_str::<ConnectionsFile>(&content) else {
        return HashMap::new();
    };
    if file.version != FILE_VERSION {
        return HashMap::new();
    }
    let mut records = HashMap::new();
    for (id, record) in file.connections {
        if id == record.connection_id && record_valid(&record) {
            records.insert(id, record);
        }
    }
    records
}

/// Write the file atomically (temp + rename, mode 0600).
fn write_records(
    path: &Path,
    records: &HashMap<String, McpConnectionRecord>,
) -> Result<(), anyhow::Error> {
    let ordered: std::collections::BTreeMap<&String, &McpConnectionRecord> =
        records.iter().collect();
    let doc = serde_json::json!({
        "version": FILE_VERSION,
        "connections": ordered,
    });
    atomic_write(path, &serde_json::to_string_pretty(&doc)?)?;
    Ok(())
}

/// A fresh pending record for one connection (verification fills the rest).
pub fn new_pending_record(
    connection_id: &str,
    service_id: &str,
    label: &str,
    endpoint: &str,
) -> McpConnectionRecord {
    let now = now_ms();
    McpConnectionRecord {
        connection_id: connection_id.to_string(),
        service_id: service_id.to_string(),
        endpoint: endpoint.to_string(),
        label: label.to_string(),
        status: McpConnectionStatus::Pending,
        created_at: now,
        updated_at: now,
        verified_at: None,
        tool_count: None,
        last_error: None,
        attempt_id: None,
    }
}
